src/lib/parser.rs

```diff
@@ -0,0 +1,217 @@
+
+use ast::{Rule, LexAST};
+
+pub struct LexParser {
+    src: String,
+    newlines: Vec<usize>,
+    ast: LexAST
+}
+
+
+impl LexParser {
+    fn new(src: String) -> LexParser {
+        LexParser{
+            src     : src,
+            newlines: vec![0],
+            ast     : LexAST::new()
+        }
+    }
+
+    fn mk_error(&self, k: LexErrorKind, off: usize) -> LexBuildError {
+        let (line, col) = self.off_to_line_col(off);
+        LexBuildError{kind: k, line: line, col: col}
+    }
+
+    fn off_to_line_col(&self, off: usize) -> (usize, usize) {
+        if off == self.src.len() {
+            let line_off = *self.newlines.iter().last().unwrap();
+            return (self.newlines.len(), self.src[line_off..].chars().count() + 1);
+        }
+        let (line_m1, &line_off) = self.newlines.iter()
+                                                .enumerate()
+                                                .rev()
+                                                .find(|&(_, &line_off)| line_off <= off)
+                                                .unwrap();
+        let c_off = self.src[line_off..]
+                        .char_indices()
+                        .position(|(c_off, _)| c_off == off - line_off)
+                        .unwrap();
+        return (line_m1 + 1, c_off + 1);
+    }
+
+    fn parse(&mut self) -> LexBuildResult<usize> {
+        let mut i = try!(self.parse_declarations(0));
+        i = try!(self.parse_rules(i));
+        // We don't currently support the subroutines part of a specification. One day we might...
+        match self.lookahead_is("%%", i) {
+            Some(j) => {
+                if try!(self.parse_ws(j)) == self.src.len() { Ok(i) }
+                else {
+                    Err(self.mk_error(LexErrorKind::RoutinesNotSupported, i))
+                }
+            }
+            None    => Ok(i)
+        }
+    }
+
+    fn parse_declarations(&mut self, mut i: usize) -> LexBuildResult<usize> {
+        i = try!(self.parse_ws(i));
+        if self.lookahead_is("%%", i).is_some() { return Ok(i + 2); }
+        if i < self.src.len() {
+            Err(self.mk_error(LexErrorKind::UnknownDeclaration, i))
+        }
+        else {
+            Err(self.mk_error(LexErrorKind::PrematureEnd, i - 1))
+        }
+    }
+
+    fn parse_rules(&mut self, mut i: usize) -> LexBuildResult<usize> {
+        loop {
+            i = try!(self.parse_ws(i));
+            if i == self.src.len() { break; }
+            if self.lookahead_is("%%", i).is_some() { break; }
+            i = try!(self.parse_rule(i));
+        }
+        Ok(i)
+    }
+
+    fn parse_rule(&mut self, i: usize) -> LexBuildResult<usize> {
+        let line_len = self.src[i..]
+                       .chars()
+                       .take_while(|&c| c != '\n')
+                       .fold(0, |a, c| a + c.len_utf8());
+        let line     = self.src[i..i + line_len].trim_right();
+        let rspace   = match line.rfind(' ') {
```

Comment:
OK, this one thoroughly confused me ;) It's actually an incorrect test, because `\n` is expanded in the test. So you're actually inputting this lex file:

```
"
%%
[
    ]+ ws
        "
```
which is, of course, invalid.

Now that you've accidentally pointed this out, I will fix the `\t` in a couple of the existing tests (which is, as it so happens, correct, but not for the reasons I thought it was!).